### PR TITLE
Feature/seab 3656/topic sentence

### PR DESCRIPTION
**Pull request by garyluu**
**Description**
Add topic sentence display to workflow information card. Also add the ability to update it (like workflow path).

Forgot to mention...it's slightly different than existing info tab buttons and such because it's closer to mocks. Presumably the rest of the info tab is converted to the mock ups before the next release.

Soon to be replaced by phase 2

**Issue**
For SEAB-3656

### CODE REVIEW

File: src/app/container/info-tab/info-tab.component.html [hunk 1]
```diff
@@ -243,6 +243,49 @@
...
+                <input
+                  *ngIf="topicEditing"
+                  minlength="3"
+                  maxlength="256"
+                  type="text"
+                  name="topic"
+                  class="w-100"
+                  data-cy="topicInput"
+                  [(ngModel)]="tool.topic"
+                  placeholder="Short description of the workflow"
```
Comment (coverbeck) [inline src/app/container/info-tab/info-tab.component.html:261]: workflow -> tool

File: src/app/container/info-tab/info-tab.service.ts [hunk 2]
```diff
@@ -85,6 +86,29 @@ export class InfoTabService extends Base {
...
   }
 
+  setTopicEditing(editing: boolean) {
+    this.topicEditing$.next(editing);
+  }
+
+  saveTopic(tool: DockstoreTool, callback: () => void) {
+    this.alertService.start('Updating topic');
+    const partialTool = this.getPartialToolForUpdate(tool);
+    console.log(tool.topic);
```
Comment (coverbeck) [inline src/app/container/info-tab/info-tab.service.ts:96]: Suggest removing.

File: src/app/container/info-tab/info-tab.service.ts [hunk 3]
```diff
@@ -85,6 +86,29 @@ export class InfoTabService extends Base {
...
+    this.containersService.updateContainer(this.tool.id, partialTool).subscribe(
+      (response) => {
+        this.alertService.detailedSuccess();
+
+        const newTopic = response.topic;
+        console.log(newTopic);
```
Comment (coverbeck) [inline src/app/container/info-tab/info-tab.service.ts:102]: I'd remove.

File: src/app/container/info-tab/info-tab.service.ts [hunk 4]
```diff
@@ -85,6 +86,29 @@ export class InfoTabService extends Base {
     this.wdlTestPathEditing$.next(editing);
   }
 
+  setTopicEditing(editing: boolean) {
+    this.topicEditing$.next(editing);
+  }
+
+  saveTopic(tool: DockstoreTool, callback: () => void) {
```
Comment (coverbeck) [inline src/app/container/info-tab/info-tab.service.ts:93]: Suggest renaming `callback` to something like `errorCallback`.

File: src/app/workflow/info-tab/info-tab.service.ts [hunk 5]
```diff
@@ -80,6 +85,26 @@ export class InfoTabService {
     this.forumUrlEditing$.next(editing);
   }
 
+  /**
+   * Warning, this could potentially update a few other properties
+   * @param workflow
+   */
+  saveTopic(workflow: Workflow, callback: () => void) {
```
Comment (coverbeck) [inline src/app/workflow/info-tab/info-tab.service.ts:92]: Same suggestion on renaming `callback`.

**Review by coverbeck — APPROVED** [review]

**Review by NatalieEO — APPROVED** [review]

**Comment by denis-yuen** [comment]
> Forgot to mention...it's slightly different than existing info tab buttons and such because it's closer to mocks. Presumably the rest of the info tab is converted to the mock ups before the next release. 

:+1:

**Comment by denis-yuen** [comment]
> Also add the ability to update it (like workflow path). 

Huh, interesting. Missed this in the webservice, who wins if it gets updated in Dockstore and on GitHub?

**Review by denis-yuen — COMMENTED** [review]
Display looks good. 
Unsure about editing

**Comment by garyluu** [comment]
> Display looks good. Unsure about editing (as opposed to update on refresh + github app events)

Refresh doesn't automatically set it because it potentially wipes out some other source that the user manually sets. So that shouldn't be done at all (unless it's for a new workflow which it currently does).

GitHub apps events are version level, so it generally should not be affecting workflow level information.  It also has the disadvantage of only affecting GitHub Apps workflows so we need some way to set it for new non-GitHub Apps workflows (otherwise it would pull from GitHub repo and the user potentially doesn't want it)

**Comment by denis-yuen** [comment]
> Refresh doesn't automatically set it because it potentially wipes out some other source that the user manually sets. So that shouldn't be done at all (unless it's for a new workflow which it currently does). 

That's kinda why I'm thinking there shouldn't be any other source, manual or otherwise. It simplifies things if GitHub is the source of truth. Otherwise, we end up with topic sentences being up-to-date after migration but quickly falling out of date if people don't notice they have to edit themselves after editing the topic on GitHub.

> GitHub apps events are version level, so it generally should not be affecting workflow level information. 

I think this is ok. We have `publish` and  `latestTagAsDefault` already

> It also has the disadvantage of only affecting GitHub Apps workflows so we need some way to set it for new non-GitHub Apps workflows  

This would be refresh or if we build out "bitbucket apps" someday (unlikely though based on usage).

**Comment by garyluu** [comment]
> > Refresh doesn't automatically set it because it potentially wipes out some other source that the user manually sets. So that shouldn't be done at all (unless it's for a new workflow which it currently does).
> 
> That's kinda why I'm thinking there shouldn't be any other source, manual or otherwise. It simplifies things if GitHub is the source of truth. Otherwise, we end up with topic sentences being up-to-date after migration but quickly falling out of date if people don't notice they have to edit themselves after editing the topic on GitHub.
> 
> > GitHub apps events are version level, so it generally should not be affecting workflow level information.
> 
> I think this is ok. We have `publish` and `latestTagAsDefault` already
> 
> > It also has the disadvantage of only affecting GitHub Apps workflows so we need some way to set it for new non-GitHub Apps workflows
> 
> This would be refresh or if we build out "bitbucket apps" someday (unlikely though based on usage).

GitHub Apps can't act as the source of truth for updates because once the admin endpoint that pulls info from the GitHub repo's description is ran, half of the workflows and all of the tools on Dockstore will potentially have unwanted topics with no way to change it. 

So for simplicity, I agree that there should ideally be one source for updates (the update endpoint) which works for all types of workflows (hosted, github apps, other) and tools. The topic shouldn't change often enough for it to be a hassle. Also changing the topic on the site is faster than making a new commit with a modified .dockstore.yml and merging it to develop (to prevent the below)

GitHub Apps topic (even as an alternative method) suffers from the same issue as publish. Two developers are making changes to the same workflow on GitHub on different branches. Dockstore just bounces back and forth between different publish and topic states.

Just like how GitHub Apps has the publish feature, the Dockstore website also has a publish button which would mirror topics if we decided to add topic to the .dockstore.yml.

I think it's one thing to favor GitHub Apps by adding more features to it, but it's another thing to cripple all the other non-GitHub Apps workflows to do so.

**Review by wshands — APPROVED** [review]

**Comment by garyluu** [comment]
This has major problems with workflows (which was implemented with existing bootstrap classes)

**Comment by garyluu** [comment]
Changed a few things. Truncate on two lines in search. Made workflow info tab component not use bootstrap which fixes the wrapping issue

**Review by NatalieEO — APPROVED** [review]

**Review by coverbeck — COMMENTED** [review]
Does this need to be updated given our discussions about the backend storing both the GitHub description and the user's override?

**Pull request merged**